conn_browser.go

```diff
@@ -0,0 +1,292 @@
+// +build js,wasm
+
+package websocket
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+	"sync"
+	"syscall/js"
+	"time"
+)
+
+const (
+	webSocketStateConnecting = 0
+	webSocketStateOpen       = 1
+	webSocketStateClosing    = 2
+	webSocketStateClosed     = 3
+)
+
+var errConnectionClosed = errors.New("connection is closed")
+
+// Conn implements net.Conn interface for WebSockets in js/wasm.
+type Conn struct {
+	js.Value
+	messageHandler *js.Func
+	closeHandler   *js.Func
+	errorHandler   *js.Func
+	mut            sync.Mutex
+	currDataMut    sync.RWMutex
+	currData       bytes.Buffer
+	closeOnce      sync.Once
+	closeSignal    chan struct{}
+	dataSignal     chan struct{}
+	localAddr      net.Addr
+	remoteAddr     net.Addr
+	firstErr       error
+}
+
+// NewConn creates a Conn given a regular js/wasm WebSocket Conn.
+func NewConn(raw js.Value) *Conn {
+	conn := &Conn{
+		Value:       raw,
+		closeSignal: make(chan struct{}),
+		dataSignal:  make(chan struct{}, 1),
+		localAddr:   NewAddr("0.0.0.0:0"),
+		remoteAddr:  getRemoteAddr(raw),
+	}
+	// Force the JavaScript WebSockets API to use the ArrayBuffer type for
+	// incoming messages instead of the Blob type. This is better for us because
+	// ArrayBuffer can be converted to []byte synchronously but Blob cannot.
+	conn.Set("binaryType", "arraybuffer")
+	conn.setUpHandlers()
+	return conn
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	if err := c.checkOpen(); err != nil {
+		return c.readAfterErr(b)
+	}
+
+	for {
+		c.currDataMut.RLock()
+		n, err := c.currData.Read(b)
+		c.currDataMut.RUnlock()
+		if err != nil && err != io.EOF {
+			// Return any unexpected errors immediately.
+			return n, err
+		} else if n == 0 || err == io.EOF {
+			// There is no data ready to be read. Wait for more data or for the
+			// connection to be closed.
+			select {
+			case <-c.dataSignal:
+				continue
+			case <-c.closeSignal:
+				return c.readAfterErr(b)
+			}
+		} else {
+			return n, err
+		}
+	}
+}
+
+// readAfterError reads from c.currData. If there is no more data left it
+// returns c.firstErr if non-nil and otherwise returns io.EOF.
+func (c *Conn) readAfterErr(b []byte) (int, error) {
+	c.currDataMut.RLock()
+	n, err := c.currData.Read(b)
+	c.currDataMut.RUnlock()
+	if n == 0 {
+		if c.firstErr != nil {
+			return 0, c.firstErr
```

Comment:
What's `firstErr`? If there's a connection _error_ (not just a polite close), we should abort immediately without bothering to read the rest of the buffer.

Reply:
@Stebalien Apparently, due to security concerns [most browsers always use `code 1006: connection unexpectedly closed`](https://stackoverflow.com/questions/19304157/getting-the-reason-why-websockets-closed-with-close-code-1006) whenever a WebSocket connection is closed. I don't think there is any way for us to differentiate between a polite close and an unexpected error in the underlying WebSocket connection.

Given that constraint, what do you want to do here?

Reply:
I'm not seeing anything to suggest that. From the looks of it, 1006 means "something went wrong". We just need to distinguish between that case and "the connection was closed correctly".

Reply:
Ah. I see https://stackoverflow.com/a/53340067. That's a bug in chrome and there's nothing we can do about it. Where are you seeing the talk about this being a security feature?

Reply:
From the link I shared above:

> However, Chrome will rarely report any close code 1006 reasons to the javascript side. This is likely due to client security rules in the WebSocket spec to prevent abusing websocket. (such as using it to scan for open ports on a destination server, or for generating lots of connections for a denial-of-service attack).

Empirically, `1006` is the only error code I've observed from Chrome.

Reply:
Hm.. After reading this again, the quote doesn't exactly back up what I said. It just says that *if* there is a code `1006` error there is no way to get additional information. Let me do some more research. Maybe the reason I've been seeing `1006` has to do with how the connection is closed.

Reply:
After doing some more testing, I have observed Chrome reporting status code `1005` (also with an empty `reason` field). However, this doesn't really help us much. I can't find any way to produce code `1000` which is what we would expect for a polite close. This might be because of the bug in Chrome that you mentioned or because `gorilla/websocket` isn't sending the appropriate close frame?

There is that comment you shared from StackOverflow that claims Chrome never reports code `1000` but I did find [these tests in Chromium](https://github.com/crosswalk-project/chromium-crosswalk/blob/dd39e4fbbc1e4417cef2f908cc0a42af1e2120aa/third_party/WebKit/LayoutTests/http/tests/websocket/close-code-and-reason-expected.txt) which suggest there are at least some cases where it does report code `1000`.

How should we proceed?

Reply:
Thanks for looking into this. Turns out, we _aren't_ sending a nice close message: https://github.com/libp2p/go-ws-transport/pull/54

Reply:
> It sounds like maybe you aren't satisfied with this approach?

I'm fine with an async approach. The network itself is async so there's nothing we can do about that.

My concern is:

1. When we receive an _error_, we should start failing reads/writes ASAP. This doesn't need to be synchronous, I just don't want to bother finishing reading.
2. When we receive an EOF (nice close), we should continue reading to the end.

Note: We're mostly getting into pedantic correctness edge-cases here.